### PR TITLE
[concurrency] quick-and-dirty implementation of MainActor in the runtime system

### PR DESCRIPTION
**Pull request by kavon**
To help unblock work that depends on the `@MainActor`, here's a hacky implementation of MainActor. This implementation should be replaced once we have better support for custom executors in the runtime system.

TODOs:
- [x] Have code marked with `@MainActor` actually run on the main thread.
- [x] Fix deadlock in `test/Concurrency/Runtime/mainactor.swift`. That test blocks the main thread while waiting on a task that needs the main thread.
- [x] ~Create working regression test for Linux.~ Defer support on Linux until PR #35215 is merged

### CODE REVIEW

**Comment by kavon** [comment]
Here's a simplified version of `mainactor.swift`:

```
@MainActor
func helloMainActor() {}

runAsyncAndBlock {
  let handle = Task.runDetached(operation: { await helloMainActor() })
  do {
    try await handle.get()
  } catch {}
}
```

A correct implementation of the MainActor attribute would deadlock in that code, and it does currently:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff204cd8e2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff204ffe6f libsystem_pthread.dylib`_pthread_cond_wait + 1254
    frame #2: 0x0000000100141bd9 libswift_Concurrency.dylib`swift::ConditionPlatformHelper::wait(condition=<unavailable>, mutex=<unavailable>) at MutexPThread.cpp:88:3 [opt]
    frame #3: 0x000000010013f6bb libswift_Concurrency.dylib`swift_task_runAndBlockThread [inlined] swift::ConditionVariable::Mutex::wait(this=<unavailable>, condition=<unavailable>) at Mutex.h:198:7 [opt]
    frame #4: 0x000000010013f6a5 libswift_Concurrency.dylib`swift_task_runAndBlockThread [inlined] void swift::ConditionVariable::Mutex::withLockOrWait<(anonymous namespace)::RunAndBlockSemaphore::wait()::'lambda'()>(this=<unavailable>)::RunAndBlockSemaphore::wait()::'lambda'())::'lambda'()::operator()() const at Mutex.h:248 [opt]
    frame #5: 0x000000010013f698 libswift_Concurrency.dylib`swift_task_runAndBlockThread [inlined] decltype(this=<unavailable>, criticalSection=(anonymous class) @ 0x00007fce9e5b1350)) swift::ConditionVariable::Mutex::withLock<void swift::ConditionVariable::Mutex::withLockOrWait<(anonymous namespace)::RunAndBlockSemaphore::wait()::'lambda'()>(swift::ConditionVariable&, (anonymous namespace)::RunAndBlockSemaphore::wait()::'lambda'())::'lambda'()>((anonymous namespace)::RunAndBlockSemaphore::wait()::'lambda'()) at Mutex.h:217 [opt]
    frame #6: 0x000000010013f690 libswift_Concurrency.dylib`swift_task_runAndBlockThread [inlined] void swift::ConditionVariable::Mutex::withLockOrWait<(anonymous namespace)::RunAndBlockSemaphore::wait()::'lambda'()>(this=<unavailable>, condition=<unavailable>, criticalSection=<unavailable>)::RunAndBlockSemaphore::wait()::'lambda'()) at Mutex.h:246 [opt]
    frame #7: 0x000000010013f690 libswift_Concurrency.dylib`swift_task_runAndBlockThread [inlined] (anonymous namespace)::RunAndBlockSemaphore::wait(this=<unavailable>) at Task.cpp:370 [opt]
    frame #8: 0x000000010013f690 libswift_Concurrency.dylib`swift_task_runAndBlockThread(function=<unavailable>, functionContext=0x0000000000000000) at Task.cpp:463 [opt]
    frame #9: 0x0000000100002495 a.out`main + 37
    frame #10: 0x00007fff2051a621 libdyld.dylib`start + 1
(lldb) 
```

So that's good 😆 

I need a working `yield` operation so that `runAndBlockThread` in that test doesn't prevent the detached task from progressing.

**Comment by DougGregor** [comment]
I think the deadlock can be addressed by using `https://developer.apple.com/documentation/dispatch/1452860-dispatchmain` instead of `runAsyncAndBlock`, then exit from the main task.

**Comment by kavon** [comment]
Ok so I've fixed the deadlock with `dispatchMain`, but `Thread.isMainThread` returns false, even though the debugger appears to say it's running on the main queue:

```
Process 8627 stopped
* thread #2, queue = 'com.apple.main-thread', stop reason = instruction step into
    frame #0: 0x0000000100002b54 a.out`main.helloMainActor() -> Swift.Never + 20
a.out`main.helloMainActor() -> Swift.Never:
    0x100002b4d <+13>: movq   0x5b5c(%rip), %rdi        ; (void *)0x00007fff88476170: NSThread
->  0x100002b54 <+20>: callq  0x100006f54               ; symbol stub for: swift_getInitializedObjCClass
    0x100002b59 <+25>: movq   0x5a38(%rip), %rsi        ; "isMainThread"
    0x100002b60 <+32>: movq   %rax, %rdi
    0x100002b63 <+35>: callq  0x100006ebe               ; symbol stub for: objc_msgSend
Target 0: (a.out) stopped.
```

Perhaps the test is not correct here.

**Comment by kavon** [comment]
Turns out the test was not correct, but I've now fixed it.

**Comment by kavon** [comment]
@swift-ci please test

File: stdlib/public/Concurrency/GlobalExecutor.cpp [hunk 1]
```diff
@@ -167,3 +176,25 @@ void swift::swift_task_enqueueGlobal(Job *job) {
...
 }
+
+
+/// Enqueues a task on the main executor.
+/// FIXME: only exists for the quick-and-dirty @MainActor implementation.
+void swift::swift_task_enqueueMainExecutor(Job *job) {
+  assert(job && "no job provided");
+
+#if SWIFT_CONCURRENCY_COOPERATIVE_GLOBAL_EXECUTOR
+  insertIntoJobQueue(job);
```
Comment (kavon) [inline stdlib/public/Concurrency/GlobalExecutor.cpp:187]: If dispatch isn't available, we'll hit this bit of incorrect code. Should I get it working in this case too?

**Comment by kavon** [comment]
@swift-ci please clean test macOS platform

**Comment by kavon** [comment]
@swift-ci please test macOS platform

**Comment by kavon** [comment]
@swift-ci please smoke test Linux platform

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by kavon** [comment]
I revised the test a bit so that it does not need an explicit call to `CFRunLoopRun`, since #35215 will enable async main. So, this PR is currently awaiting on that PR to land.

**Comment by kavon** [comment]
@swift-ci please test

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Pull request merged**